crates/adapters/src/integrated/postgres/output.rs

```diff
@@ -0,0 +1,299 @@
+        config: &PostgresWriterConfig,
+        key_schema: &Option<Relation>,
+        _schema: &Relation,
+        controller: Weak<ControllerInner>,
+    ) -> Result<Self, ControllerError> {
+        let table = config.table.to_owned();
+
+        let config = postgres::Config::from_str(&config.uri).map_err(|e| {
+            ControllerError::output_transport_error(
+                endpoint_name,
+                true,
+                anyhow!("error parsing postgres connection string: {e}"),
+            )
+        })?;
+
+        let mut client = config.connect(NoTls).map_err(|e| {
+            ControllerError::output_transport_error(
+                endpoint_name,
+                true,
+                anyhow!("failed to connect to postgres: {e}"),
+            )
+        })?;
+
+        let keys: HashSet<String> = key_schema
+            .as_ref()
+            .map(|sch| sch.fields.iter().map(|f| f.name.to_string()).collect())
+            .ok_or(ControllerError::not_supported(
+                "postgres output connector requires index or primary key definition",
+            ))?;
+
+        let insert = {
+            let insert = format!(
+                "INSERT INTO {table} SELECT * FROM jsonb_populate_recordset(NULL::{table}, $1::jsonb)"
+            );
+            client
+                .prepare_typed(&insert, &[postgres::types::Type::VARCHAR])
+                .unwrap_or_else(|e| panic!("failed to prepare insert statement: `{insert}`: {e}"))
+        };
+
+        let delete = {
+            let (table_keys, d_keys): (Vec<_>, Vec<_>) = keys
+                .iter()
+                .map(|k| (format!("{table}.{k}"), format!("d.{k}")))
+                .unzip();
+
+            let delete = format!(
+                "DELETE FROM {table} USING (SELECT {} FROM jsonb_populate_recordset(NULL::{table}, $1::jsonb)) as d where ({}) = ({})",
+                keys.iter().map(|k| k.as_str()).collect::<Vec<_>>().join(", "),
+                table_keys.join(", "),
+                d_keys.join(", "),
+
+            );
+            client
+                .prepare_typed(&delete, &[postgres::types::Type::VARCHAR])
+                .unwrap_or_else(|e| panic!("failed to prepare delete statement: `{delete}`: {e}"))
+        };
+
+        let out = Self {
+            endpoint_id,
+            endpoint_name: endpoint_name.to_owned(),
+            table,
+            start: Instant::now(),
+            client,
+            transaction: None,
+            insert,
+            delete,
+            keys,
+            _pin: PhantomPinned,
+        };
+
+        let _guard = out.span();
+
+        Ok(out)
+    }
+
+    fn transaction(&mut self) -> AnyResult<&mut postgres::Transaction<'static>> {
+        self.transaction.as_mut().ok_or(anyhow!("postgres: unreachable: attempting to perform a transaction that hasn't been created yet"))
+    }
+
+    pub fn insert(&mut self, value: &str) -> AnyResult<()> {
+        if value.len() <= 2 {
```

Comment:
what is this for?

Reply:
If the buffer is just `[]`, that means there's nothing to insert, and there is no need to execute this statement in the transaction.